src/config_entries.c

```diff
@@ -11,6 +11,7 @@ typedef struct config_entry_list {
 	struct config_entry_list *next;
 	struct config_entry_list *last;
 	git_config_entry *entry;
+	bool first;
```

Comment:
Sorry, I'm having a hard time understanding your good-looking commit message, so I'll try to rephrase it :

```
Whenever adding a configuration entry to the config entries structure,
we allocate two list heads.

- The first list head is added to the global list of config entries, to be
able to iterate over configuration entries while preserving their order.

- The second list head is added to the map of entries. When iterating,
an already existing entry would be added to this head, while a new
one would be directly added to the map.
```

I'm confused by "its" in your version, so I might be misreading, and the difference is meaningful ^^.

```
We can completely get rid of this secondary list by just adding a
`first` field to the first configuration entry list.
```

"`first` field to the list structure itself." I think reads clearer, "first configuration entry list" sounds like "first list head" above.

I'd split the "solution" and "optimization" ¶'s as well. Thanks for the read 😉.